woocommerce/woocommerce-ios
# [Payments] Do not display error modal when merchant cancels reader connection

Merge after #11080 
Closes: #11209



## Description
This PR suppresses "System unexpectedly cancelled" error presented via a modal view to a merchant, when the card reader discovery or connection process is manually cancelled by taping the "cancel" button.

The "System unexpectedly cancelled" modal view would appear because the `ControllerState` state would change to be `.cancel` state after a merchant taps the cancel button. This would trigger an error within reader search, which is intercepted in the `onError` completion handler. This switched the `ControllerState` to `discoveryFailed(_)` immediately, which triggers `CardReaderConnectionController.onDiscoveryFailed` that presents the error in a modal to the merchant.

## Changes
We add additional logic when an error during reader search is captured, if there's an error but we're in the `.cancel` state:
* Track a `card_reader_connection_failed` error, rather than `card_reader_discovery_failed`, since discovery either has not completed when cancelled, or was completed but ultimately not connected the candidate reader.
* Pass and track `.connectionError` as the `CancellationSource`. While we're already in the `.cancel` state at this point, the error captured is unknown, so we declare it explicitly.

## Testing instructions
1. With a disconnected reader
2. Go to Menu > Payments > Manage card reader > Tap Connect Card Reader > Tap cancel before the discovery finishes
3. Observe no "System unexpectedly cancelled" error are shown
4. Repeat the discovery until a candidate reader is found, you'll see the "Do you want to connect to reader XYZ" screen
5. Tap the cancel button
6. Observe no "System unexpectedly cancelled" error are shown

![2023-11-21 at 12 47 12 supress cancel error ](https://github.com/woocommerce/woocommerce-ios/assets/3812076/2938fdc1-d50d-4bf2-bd54-2f85a5108248)
